providers/alibaba/src/airflow/providers/alibaba/cloud/log/oss_task_handler.py

```
@@ -95,6 +95,9 @@ def read(self, relative_path, ti: RuntimeTI) -> tuple[LogSourceInfo, LogMessages
             return messages, logs
         return messages, None
 
+    def stream(self, relative_path: str, ti: RuntimeTI) -> LogResponse:
+        raise NotImplementedError
```

Hmmm, it's a bit annoying that we need to define this in all the subclasses (as otherwise it fails the protocol check).

I wonder if there's a different option we could take here 🤔

---

> Hmmm, it's a bit annoying that we need to define this in all the subclasses (as otherwise it fails the protocol check).

So do I

> I wonder if there's a different option we could take here 🤔

I had though about another approach by replacing the `RemoteLogIO` Protocal with ABC, then the `stream` method will raise `NotImplementedError` by default. No sure if there is any concern of using ABC instead of Protocal ?

---

Other option would be to have two protocols (one with just read, and one which subclasses it that adds in stream), and then either mark it as "runtime checkable", or simply do something like this:

```python
        logs: LogMessages | list[RawLogStream] | None  # extra typing to void mypy assignment error
        if stream := getattr(remote_io, "stream"):
            # Use .stream interface if provider's RemoteIO supports it
            sources, logs = stream(path, ti)
            return sources, logs or []
        else:
            # Fallback to .read interface
            sources, logs = remote_io.read(path, ti)
            return sources, logs or []
```

(This also has the benefit of not masking potential exceptions from inside the stream function itself -- for instance if the stream function was implemented but raised an AttributeError inside it then it would be lost and the read fallback path chosen instead)

---

I just added another `RemoteLogStreamIO` Protocal and having `REMOTE_TASK_LOG` annotated as `REMOTE_TASK_LOG: RemoteLogIO | RemoteLogStreamIO | None = None`.
Additionally, mypy is not waring and the CI have passed.

---

Ready for re-review then?

---

Yes, and the CI also passes after I mark it for review.

---

Given we now do `getattr(remote_io, "stream")` we don't want this method here anymore do we, as we don't want this to be called.